float8_experimental/float8_dynamic_linear.py

```diff
@@ -142,12 +147,34 @@ def from_float(
 
 
 class Float8DynamicLinearWeightTensor(torch.Tensor):
-    # TODO: Remove `module` arg, save state on subclass, and propagate it.
-    def fsdp_pre_all_gather(
-        self, module: nn.Module
```

Comment:
The problem statement is that in order to implement the pre-all-gather transform, the subclass needs some state (e.g. mainly emulate but preferably the cast function as well). In the first prototype, I shortcutted by passing `module` into `fsdp_pre_all_gather()` so that the transform could read state off the module instead of storing it on the subclass itself.

However, the morally right thing (from a design perspective) should be to put that state on the subclass. I wonder though, how does this kind of  `__torch_function__` subclass interact with `torch.compile` today. cc: @bdhirsh